beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -160,6 +161,47 @@ func (s *Service) processPendingBlocks(ctx context.Context) error {
 			default:
 			}
 
+			if b.Block().Version() >= version.Deneb {
+				commitments, err := b.Block().Body().BlobKzgCommitments()
+				if err != nil {
+					return err
+				}
+				// If the block has kzg commitments, block until all the commitments are received.
+				if len(commitments) > 0 {
+					cp := s.cfg.chain.FinalizedCheckpt()
+					// Choose the best peer to request blob sidecars.
+					_, bestPeers := s.cfg.p2p.Peers().BestFinalized(maxPeerRequest, cp.Epoch)
+					if len(bestPeers) == 0 {
+						return errors.New("no peers to request blob sidecars")
+					}
+
+					// Build request for blob sidecars.
+					ids := make([]*eth.BlobIdentifier, len(commitments))
+					for i := range commitments {
+						ids[i] = &eth.BlobIdentifier{Index: uint64(i), BlockRoot: blkRoot[:]}
+					}
+					req := p2ptypes.BlobSidecarsByRootReq(ids)
+
+					// Get the context byte versions from the genesis validator root.
+					ctxByte, err := ContextByteVersionsForValRoot(s.cfg.chain.GenesisValidatorsRoot())
+					if err != nil {
+						return err
+					}
+
+					// Send the request to the best peer.
+					blobSidecars, err := SendBlobSidecarByRoot(ctx, s.cfg.clock, s.cfg.p2p, bestPeers[0], ctxByte, &req)
+					if err != nil {
+						return err
+					}
+
+					// Save the blob sidecars to the database to be verified later.
+					err = s.cfg.beaconDB.SaveBlobSidecar(ctx, blobSidecars)
+					if err != nil {
+						return err
+					}
+				}
+			}
+
```

Comment:
Does this need to account for throttling / rate limiting? I'm not familiar with how `processPendingBlocks` works - if this fails, will it be put back into a retry queue? Do we need to check if the slot is within the retention window? Do we need to consider rate limiting and peer scoring (beyond what we get from `BestFinalized`).

Reply:
> if this fails, will it be put back into a retry queue? 

Yep, the pending queue will retry itself every 4s

> Do we need to check if the slot is within the retention window?

Given this is by root, I'm inclined to say no, but I could be wrong here

> Do we need to consider rate limiting and peer scoring (beyond what we get from BestFinalized)?

This addition mirrors the get pending block by root implementation which doesn't quite have rate limiting and peer scoring